File: TSOS/C#/lab_4/DSP.Lab4.Api/Butterfly.cs
using System.Numerics;
using System;

namespace DSP.Lab4.Api
{
    public static class Butterfly
    {
        public const double SinglePi = Math.PI;
        public const double DoublePi = 2 * Math.PI;

        public static Complex[] DecimationInTime(Complex[] frame, bool direct)
        {
            if (frame.Length == 1) return frame;
            int frameHalfSize = frame.Length >> 1;   // frame.Length/2
            int frameFullSize = frame.Length;

            Complex[] frameOdd = new Complex[frameHalfSize];
            Complex[] frameEven = new Complex[frameHalfSize];
            for (int i = 0; i < frameHalfSize; i++)
            {
                int j = i << 1;               // i = 2*j;
                frameOdd[i] = frame[j + 1];
                frameEven[i] = frame[j];
            }

            Complex[] spectrumOdd = DecimationInTime(frameOdd, direct);
            Complex[] spectrumEven = DecimationInTime(frameEven, direct);

            double arg = direct 
                ? -DoublePi / frameFullSize 
                : DoublePi / frameFullSize;
            Complex omegaPowBase = new Complex(Math.Cos(arg), Math.Sin(arg));
            Complex omega = Complex.One;
            Complex[] spectrum = new Complex[frameFullSize];

            for (int j = 0; j < frameHalfSize; j++)
            {
                spectrum[j] = spectrumEven[j] + omega * spectrumOdd[j];
                spectrum[j + frameHalfSize] = spectrumEven[j] - omega * spectrumOdd[j];
                omega *= omegaPowBase;
            }

            return spectrum;
        }
    }
}

File: TSOS/JS/lab_4/src/lab/Butterfly.ts
import * as math from 'mathjs';

const DoublePi = 2 * Math.PI;
export class Butterfly {
  public static DecimationInTime(
    frame: math.Complex[],
    direct: boolean
  ): math.Complex[] {
    if (frame.length === 1) return frame;
    const frameHalfSize = frame.length >> 1; // /2
    const frameFullSize = frame.length;
    const frameOdd: math.Complex[] = new Array<math.Complex>(frameHalfSize);
    const frameEven: math.Complex[] = new Array<math.Complex>(frameHalfSize);

    for (let i = 0; i < frameHalfSize; i++) {
      let j = i << 1; // i = 2 * j;
      frameOdd[i] = frame[j + 1];
      frameEven[i] = frame[j];
    }
    const spectrumOdd: math.Complex[] = Butterfly.DecimationInTime(
      frameOdd,
      direct
    );
    const spectrumEven: math.Complex[] = Butterfly.DecimationInTime(
      frameEven,
      direct
    );
    let arg = direct ? -DoublePi / frameFullSize : DoublePi / frameFullSize;
    const omegaPowBase: math.Complex = math.complex(
      Math.cos(arg),
      Math.sin(arg)
    );
    let omega: math.Complex = math.complex(1, 0);
    const spectrum: math.Complex[] = new Array(frameFullSize);
    for (let j = 0; j < frameHalfSize; j++) {
      const temp = math.multiply(omega, spectrumOdd[j]);
      spectrum[j] = math.add(spectrumEven[j], temp) as math.Complex;
      spectrum[j + frameHalfSize] = math.subtract(
        spectrumEven[j],
        math.multiply(omega, spectrumOdd[j])
      ) as math.Complex;

      omega = math.multiply(omega, omegaPowBase) as math.Complex;
    }
    return spectrum;
  }
}
